text/0000-scoped-take-2.md

```diff
@@ -0,0 +1,261 @@
+and a new `thread::scoped` API that takes advantage of it.
+
+## Deferred computation
+
+Here's a very simple API for deferring computation:
+
+```
+// in std:
+
+mod thread {
+    pub struct Scope<'a> { ... }
+
+    pub fn scope<'a, F, R>(f: F) -> R where F: FnOnce(&Scope<'a>) -> R;
+
+    impl<'a> Scope<'a> {
+        pub fn defer<F>(&self, f: F) where F: FnOnce() + 'a;
+    }
+}
+```
+
+(This is put in the `std::thread` module because (1) scopes are per-thread
+concepts (2) it fits nicely with other functionality like `panicking` and
+`catch_panic`.)
+
+You call `scope` to introduce a new `Scope` value, which is passed into a
+closure that is immediately invoked.
+
+The closure can use the `defer` method to register callbacks to invoke upon
+*any* exit from the callback, including unwinding. Unlike RAII guards, there is
+no way to leak these callbacks, and the implementation shown below works around
+[known cases](https://github.com/rust-lang/rust/issues/14875) of destructor
+leakage.
+
+This is a generally useful mechanism that can avoid the need to create custom
+RAII guards for situations where you might use `try`/`finally` in other
+languages. But it is also just the support needed for scoped threads.
+
+## Scoped threads
+
+To recover scoped threads, we extend the `Scope` type with a method for spawning
+threads:
+
+```rust
+impl<'a> Scope<'a> {
+    pub fn spawn<F, T>(&self, f: F) -> thread::JoinHandle<T> where
+        F: FnOnce() -> T + Send + 'a,
+        T: Send + 'a;
+}
+```
+
+Like the original `thread::scoped` API, this allows for the child thread's
+closure to borrow data (with lifetime `'body`) from the parent thread. These
+borrows are bounded by the lifetime of the `Scope` value, and the implementation
+uses the `defer` method to add a callback that will join on (wait for completion
+of) the child thread on exit from the scope -- thus restoring memory safety.
+
+Note that, while previously one might return `JoinGuard`s outward to expand the
+scope of joining, the pattern here is reversed: you call `scope` at the
+outer-most scope, and then pass a reference to the `Scoped` value inward.
+
+Putting it all together, here's an example from TRPL, and a version adjust to the new API:
+
+```rust
+// using the thread::scoped API
+fn old_trpl_example() {
+    let data = Mutex::new(vec![1u32, 2, 3]);
+
+    let guards: Vec<_> = (0..2).iter().map(|_| {
+        thread::scoped(|| {
+            let mut data = data.lock().unwrap();
+            data[i] += 1;
+        })
+    }).collect();
+
+    // threads implicitly joined here, when `guards` goes out of scope and drops
+    // its contents
+}
+```
+
+```rust
+// using the proposed thread::scope API
+fn new_trpl_example() {
+    let data = Mutex::new(vec![1u32, 2, 3]);
+
+    thread::scope(|s| {
+        for i in 0..2 {
+            s.spawn(|| {
+                let mut data = data.lock().unwrap();
+                data[i] += 1;
+            });
```

Comment:
@kballard 

I'm a bit confused by this comment. My understanding here is that the variable `i` is inferred to capture *by value*, because addition works by value. Since it's a `Copy` type, that means that a copy is moved into each closure as it's created. So I don't see any data race with the example as given. Can you clarify? I'd also be curious to see the exact use of `scoped` that produced the borrowck error.